File: ga/gpt3_5_result_2024_6/result/tree_gpt/3_tree.c
use std::ptr;

struct TreeNode {
    element: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

type SearchTree = Option<Box<TreeNode>>;
type Position = Option<Box<TreeNode>>;

fn make_empty(t: SearchTree) -> SearchTree {
    if let Some(mut node) = t {
        node.left = make_empty(node.left);
        node.right = make_empty(node.right);
    }
    t
}

fn find(x: i32, t: SearchTree) -> Position {
    match t {
        None => None,
        Some(node) => {
            if x < node.element {
                find(x, node.left.clone())
            } else if x > node.element {
                find(x, node.right.clone())
            } else {
                Some(node)
            }
        }
    }
}

fn find_min(t: SearchTree) -> Position {
    match t {
        None => None,
        Some(node) => {
            if let Some(mut left) = node.left.clone() {
                find_min(Some(left))
            } else {
                Some(node)
            }
        }
    }
}

fn find_max(t: SearchTree) -> Position {
    match t {
        None => None,
        Some(mut node) => {
            while let Some(right) = node.right.clone() {
                node = right;
            }
            Some(node)
        }
    }
}

fn insert(x: i32, t: SearchTree) -> SearchTree {
    match t {
        None => Some(Box::new(TreeNode {
            element: x,
            left: None,
            right: None,
        })),
        Some(mut node) => {
            if x < node.element {
                node.left = insert(x, node.left.take());
            } else if x > node.element {
                node.right = insert(x, node.right.take());
            }
            Some(node)
        }
    }
}
 
fn delete(x: i32, t: SearchTree) -> SearchTree {
    match t {
        None => panic!("Element not found"),
        Some(mut node) => {
            if x < node.element {
                node.left = delete(x, node.left.take());
            } else if x > node.element {
                node.right = delete(x, node.right.take());
            } else {
                if let Some(mut right) = node.right.take() {
                    let min = find_min(Some(right.clone())).unwrap();
                    node.element = min.element;
                    node.right = delete(node.element, node.right.take());
                } else {
                    let tmp_node = node;
                    if node.left.is_none() {
                        node = node.right.take().unwrap();
                    } else if node.right.is_none() {
                        node = node.left.take().unwrap();
                    }
                    drop(tmp_node);
                }
            }
            Some(node)
        }
    }
}

fn retrieve(p: Position) -> i32 {
    p.unwrap().element
}

fn main() {
    let mut t: SearchTree = None;
    let mut j = 0;

    t = make_empty(t);

    for i in 0..50 {
        j = (j + 7) % 50;
        t = insert(j, t);
    }

    for i in 0..50 {
        let p = find(i, t.clone());
        if p.is_none() || retrieve(p) != i {
            println!("Error at {}", i);
        }
    }

    for i in (0..50).step_by(2) {
        t = delete(i, t);
    }

    for i in (1..50).step_by(2) {
        let p = find(i, t.clone());
        if p.is_none() || retrieve(p) != i {
            println!("Error at {}", i);
        }
    }

    for i in (0..50).step_by(2) {
        if find(i, t.clone()).is_some() {
            println!("Error at {}", i);
        }
    }

    println!(
        "Min is {}, Max is {}",
        retrieve(find_min(t.clone()).unwrap()),
        retrieve(find_max(t.clone()).unwrap())
    );
}

File: ga/4omini_result_2024_7/result/tree_mini/3_tree.rs
use std::cmp::Ord;
use std::collections::BinaryHeap;
use std::fmt::Debug;
use std::io;

#[derive(Debug)]
struct TreeNode<T> {
    element: T,
    left: Option<Box<TreeNode<T>>>,
    right: Option<Box<TreeNode<T>>>,
}

pub struct SearchTree<T> {
    root: Option<Box<TreeNode<T>>>,
}

impl<T: Ord + Copy + Debug> SearchTree<T> {
    pub fn new() -> Self {
        Self { root: None }
    }

    pub fn insert(&mut self, x: T) {
        self.root = Self::insert_rec(self.root.take(), x);
    }

    fn insert_rec(node: Option<Box<TreeNode<T>>>, x: T) -> Option<Box<TreeNode<T>>> {
        match node {
            Some(mut n) => {
                if x < n.element {
                    n.left = Self::insert_rec(n.left, x);
                } else if x > n.element {
                    n.right = Self::insert_rec(n.right, x);
                }
                Some(n)
            }
            None => Some(Box::new(TreeNode {
                element: x,
                left: None,
                right: None,
            })),
        }
    }

    pub fn delete(&mut self, x: T) {
        self.root = Self::delete_rec(self.root.take(), x);
    }

    fn delete_rec(node: Option<Box<TreeNode<T>>>, x: T) -> Option<Box<TreeNode<T>>> {
        match node {
            Some(mut n) => {
                if x < n.element {
                    n.left = Self::delete_rec(n.left, x);
                    Some(n)
                } else if x > n.element {
                    n.right = Self::delete_rec(n.right, x);
                    Some(n)
                } else {
                    // Found the node to delete
                    match (n.left.take(), n.right.take()) {
                        (None, None) => None, // No children
                        (Some(left), None) => Some(left), // Only left child
                        (None, Some(right)) => Some(right), // Only right child
                        (Some(left), Some(right)) => {
                            let min = Self::find_min(right.as_ref().unwrap());
                            n.element = min.element; // Replace with min element from right subtree
                            n.right = Self::delete_rec(Some(right), min.element);
                            Some(n)
                        }
                    }
                }
            }
            None => {
                eprintln!("Element not found");
                None
            }
        }
    }

    pub fn find(&self, x: T) -> Option<&T> {
        Self::find_rec(&self.root, x)
    }

    fn find_rec(node: &Option<Box<TreeNode<T>>>, x: T) -> Option<&T> {
        match node {
            Some(n) => {
                if x < n.element {
                    Self::find_rec(&n.left, x)
                } else if x > n.element {
                    Self::find_rec(&n.right, x)
                } else {
                    Some(&n.element)
                }
            }
            None => None,
        }
    }

    pub fn find_min(&self) -> Option<&T> {
        self.root.as_ref().and_then(Self::find_min_rec)
    }

    fn find_min_rec(node: &Box<TreeNode<T>>) -> Option<&T> {
        if let Some(left) = &node.left {
            Self::find_min_rec(left)
        } else {
            Some(&node.element)
        }
    }

    pub fn find_max(&self) -> Option<&T> {
        self.root.as_ref().and_then(Self::find_max_rec)
    }

    fn find_max_rec(node: &Box<TreeNode<T>>) -> Option<&T> {
        if let Some(right) = &node.right {
            Self::find_max_rec(right)
        } else {
            Some(&node.element)
        }
    }

    pub fn make_empty(&mut self) {
        self.root = None;
    }
}

fn main() {
    let mut tree = SearchTree::new();
    let mut j = 0;

    // Insert elements
    for i in 0..50 {
        j = (j + 7) % 50;
        tree.insert(j);
    }

    // Verify elements
    for i in 0..50 {
        if tree.find(i).is_none() {
            println!("Error at {}", i);
        }
    }

    // Delete even elements
    for i in (0..50).step_by(2) {
        tree.delete(i);
    }

    // Verify remaining odd elements
    for i in (1..50).step_by(2) {
        if tree.find(i).is_none() {
            println!("Error at {}", i);
        }
    }

    // Verify deleted even elements
    for i in (0..50).step_by(2) {
        if tree.find(i).is_some() {
            println!("Error at {}", i);
        }
    }

    // Print min and max
    if let Some(min) = tree.find_min() {
        if let Some(max) = tree.find_max() {
            println!("Min is {:?}, Max is {:?}", min, max);
        }
    }
}